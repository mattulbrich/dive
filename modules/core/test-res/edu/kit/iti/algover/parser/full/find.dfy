method Find(a: array<int>, key: int) returns (index: int)
   requires a != null
   ensures 0 <= index ==> index < a.Length && a[index] == key
   // !!! ensures index < 0 ==> forall k :: 0 <= k < a.Length ==> a[k] != key
   ensures index < 0 ==> (forall k:int :: 0 <= k && k < a.Length ==> a[k] != key)
{
   index := 0;
   while index < a.Length
      // !!! invariant 0 <= index <= a.Length
      invariant 0 <= index && index <= a.Length
      // !!! invariant forall k :: 0 <= k < index ==> a[k] != key
      invariant (forall k:int :: 0 <= k && k < index ==> a[k] != key)
   {
      if a[index] == key
      {
          // !!! return; 
      }
      index := index + 1;
   }
   index := -1;
}

