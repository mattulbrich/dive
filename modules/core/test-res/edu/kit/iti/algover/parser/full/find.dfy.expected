(COMPILATION_UNIT 
 (method Find 
  (ARGS 
   (VAR a 
    (TYPE (array int)))
   (VAR key (TYPE int))) 
  (returns 
   (VAR index (TYPE int))) 
  (requires 
   (!= a null)) 
  (ensures 
   (==> 
    (<= 0 index) 
    (&& 
     (< index 
      (Length a)) 
     (== 
      (ARRAY_ACCESS a index) key)))) 
  (ensures 
   (==> 
    (< index 0) 
    (forall k (TYPE int) 
     (==> 
      (&& 
       (<= 0 k) 
       (< k 
        (Length a))) 
      (!= 
       (ARRAY_ACCESS a k) key))))) 
  (BLOCK 
   (:= index 0) 
   (while 
    (< index 
     (Length a)) 
    (invariant 
     (&& 
      (<= 0 index) 
      (<= index 
       (Length a)))) 
    (invariant 
     (forall k (TYPE int) 
      (==> 
       (&& 
        (<= 0 k) 
        (< k index)) 
       (!= 
        (ARRAY_ACCESS a k) key)))) 
    (BLOCK 
     (if 
      (== 
       (ARRAY_ACCESS a index) key) BLOCK) 
     (:= index 
      (+ index 1)))) 
   (:= index 
    (- 1)))))
